### PR TITLE
Introduce ErrorInfoProviderOptions.ExposeExceptionDetailsMode

### LINKED ISSUE #3268: What to do with exception type name in error message?
`ExposeExceptionStackTrace = true`:
```
GraphQL.ExecutionError: Exception of type 'System.ArgumentNullException' was thrown.
 ---> System.ArgumentNullException: Exception of type 'System.ArgumentNullException' was thrown.
 ---> System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
   --- End of inner exception stack trace ---
   at GraphQL.Tests.Errors.ErrorInfoProviderTests.exposeExceptions_with_real_stack_trace() in C:\_GITHUB_\graphql-dotnet\src\GraphQL.Tests\Errors\ErrorInfoProviderTests.cs:line 166
   --- End of inner exception stack trace ---
   at GraphQL.Tests.Errors.ErrorInfoProviderTests.exposeExceptions_with_real_stack_trace() in C:\_GITHUB_\graphql-dotnet\src\GraphQL.Tests\Errors\ErrorInfoProviderTests.cs:line 170
```

`ExposeExceptionStackTrace = false`:
```
   Exception of type 'System.ArgumentNullException' was thrown.
```

We use `Exception.ToString()` when `ExposeExceptionStackTrace = true` and it appends not only stack trace in message but also prepend it with exception type like `GraphQL.ExecutionError: ` above. It worries me from time to time. In addition, I have been doubting for a long time that the stack trace should be displayed together with the error message, and not as separate field in extensions. 

For example error payload may look like in case of `ExposeExceptionStackTrace = true`
```json
{
    "errors": [
        {
            "message": "Exception of type 'System.ArgumentNullException' was thrown.",
            "extensions": {
                "stacktrace": "---> System.ArgumentNullException: Exception of type 'System.ArgumentNullException' was thrown.
 ---> System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
   --- End of inner exception stack trace ---
   at GraphQL.Tests.Errors.ErrorInfoProviderTests.exposeExceptions_with_real_stack_trace() in C:\_GITHUB_\graphql-dotnet\src\GraphQL.Tests\Errors\ErrorInfoProviderTests.cs:line 166
   --- End of inner exception stack trace ---
   at GraphQL.Tests.Errors.ErrorInfoProviderTests.exposeExceptions_with_real_stack_trace() in C:\_GITHUB_\graphql-dotnet\src\GraphQL.Tests\Errors\ErrorInfoProviderTests.cs:line 170"
                "code": ""ARGUMENT_NULL"",
                "codes": [
                    ""ARGUMENT_NULL""
                ]
            }
        }
    ]
}
```

instead of 

```json
{
    "errors": [
        {
            "message": "GraphQL.ExecutionError: Exception of type 'System.ArgumentNullException' was thrown.
 ---> System.ArgumentNullException: Exception of type 'System.ArgumentNullException' was thrown.
 ---> System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
   --- End of inner exception stack trace ---
   at GraphQL.Tests.Errors.ErrorInfoProviderTests.exposeExceptions_with_real_stack_trace() in C:\_GITHUB_\graphql-dotnet\src\GraphQL.Tests\Errors\ErrorInfoProviderTests.cs:line 166
   --- End of inner exception stack trace ---
   at GraphQL.Tests.Errors.ErrorInfoProviderTests.exposeExceptions_with_real_stack_trace() in C:\_GITHUB_\graphql-dotnet\src\GraphQL.Tests\Errors\ErrorInfoProviderTests.cs:line 170",
            "extensions": {
                "code": ""ARGUMENT_NULL"",
                "codes": [
                    ""ARGUMENT_NULL""
                ]
            }
        }
    ]
}
```

In other words I think it is worth to keep error messages "clean" without additional techical info - without stacktrace and exception type name.

### PR DESCRIPTION
**Pull request by sungam3r**
fixes #3268 

Example of `extensions.stacktrace`. Note that there are no error messages - only stacktraces and exception types.

```
GraphQL.ExecutionError:
   at GraphQL.Tests.Errors.ErrorInfoProviderTests.exposeExceptions_with_real_stack_trace_in_extensions() in C:\_GITHUB_\graphql-dotnet\src\GraphQL.Tests\Errors\ErrorInfoProviderTests.cs:line 186
[inner] System.ArgumentNullException:
   at GraphQL.Tests.Errors.ErrorInfoProviderTests.exposeExceptions_with_real_stack_trace_in_extensions() in C:\_GITHUB_\graphql-dotnet\src\GraphQL.Tests\Errors\ErrorInfoProviderTests.cs:line 182
[inner] System.ArgumentOutOfRangeException:
   <no stacktrace>
```

Or... should I just set `error.ToString()` into `extensions.stacktrace` ?

### CODE REVIEW

File: src/GraphQL/Execution/ErrorInfoProvider.cs
Comment (Shane32) [inline src/GraphQL/Execution/ErrorInfoProvider.cs:73]:
```suggestion
                        extensions.Add("stackTrace", stacktrace);
```
Names are camelCase, not lowercase, so I would think this should be "stackTrace"

File: src/GraphQL/Execution/ErrorInfoProviderOptions.cs
Comment (Shane32) [inline src/GraphQL/Execution/ErrorInfoProviderOptions.cs:19]:
```suggestion
        /// Specifies whether the extensions property, including by default the 'code', 'codes', 'data' and 'stackTrace' properties, should be serialized.
```

File: src/GraphQL/Execution/ErrorInfoProviderOptions.cs
Comment (Shane32) [inline src/GraphQL/Execution/ErrorInfoProviderOptions.cs:52]:
```suggestion
        /// Exception stack trace is located in "extensions.stackTrace" separately from exception message itself.
```

File: src/GraphQL/Execution/ErrorInfoProvider.cs
Comment (Shane32) [inline src/GraphQL/Execution/ErrorInfoProvider.cs:58]:
```suggestion
                    ? executionError.ToString()
```
I would just store `ToString()` which contains all the details, including the type name and description of each inner exception.

Example:

```
System.NullReferenceException: Outer exception message
 ---> System.InvalidOperationException: Center exception message
 ---> System.ApplicationException: Innermost exception message
   at GraphQL.MicrosoftDI.Tests.AdapterTests.AllowNullProvider() in C:\Users\Shane\source\repos\graphql-dotnet\graphql-dotnet\src\GraphQL.MicrosoftDI.Tests\AdapterTests.cs:line 26
   --- End of inner exception stack trace ---
   at GraphQL.MicrosoftDI.Tests.AdapterTests.AllowNullProvider() in C:\Users\Shane\source\repos\graphql-dotnet\graphql-dotnet\src\GraphQL.MicrosoftDI.Tests\AdapterTests.cs:line 30
   --- End of inner exception stack trace ---
   at GraphQL.MicrosoftDI.Tests.AdapterTests.AllowNullProvider() in C:\Users\Shane\source\repos\graphql-dotnet\graphql-dotnet\src\GraphQL.MicrosoftDI.Tests\AdapterTests.cs:line 35
```

Perhaps the property name should be called "details" instead of "stackTrace", however.

File: src/GraphQL/Execution/ErrorInfoProvider.cs
Comment (Shane32) [inline src/GraphQL/Execution/ErrorInfoProvider.cs:97]: This basically just replicates `ToString()` in a different manner and without the messages of exceptions (outer or inner exceptions).  I would just remove and replace with `ToString()`.

```suggestion
```

File: src/GraphQL/Execution/ErrorInfoProvider.cs
Comment (Shane32) [reply src/GraphQL/Execution/ErrorInfoProvider.cs:73]:
```suggestion
                        extensions.Add("details", stacktrace);
```
Or perhaps "details" as it will store the type names, messages and stack trace of outer and inner exceptions -- all information you'd want if enabling "stack traces".

Comment (sungam3r) [reply src/GraphQL/Execution/ErrorInfoProvider.cs:73]: So... if that piece of information is some sort of _details_ then maybe just use `exception.ToString()` and not bother with manual building formatted string?

File: src/GraphQL/Execution/ErrorInfoProvider.cs
Comment (sungam3r) [reply src/GraphQL/Execution/ErrorInfoProvider.cs:97]: Good. I would too. I just wanted to confirm this intent.

File: src/GraphQL/Execution/ErrorInfoProvider.cs
Comment (Shane32) [reply src/GraphQL/Execution/ErrorInfoProvider.cs:73]: I think that's what I suggested...  use "details" as the key and it equals `exception.ToString()`.

Comment (sungam3r) [reply src/GraphQL/Execution/ErrorInfoProvider.cs:73]: OK but then `ExposeExceptionStackTrace` is not good enough name and something like `ExposeExceptionDetails`/`ExposeExceptionDetailsMode` suits better.

Comment (sungam3r) [reply src/GraphQL/Execution/ErrorInfoProvider.cs:73]: StackTrace is just a StackTrace but we really have stack traces + exception types + exception messages

Comment (Shane32) [reply src/GraphQL/Execution/ErrorInfoProvider.cs:73]: Maybe just add `ExposeExceptionDetails` and deprecate the old property

Comment (Shane32) [reply src/GraphQL/Execution/ErrorInfoProvider.cs:73]: Or leave the old property as-is and add the new property...

**Comment by sungam3r** [comment]
Done

File: src/GraphQL.ApiTests/net6/GraphQL.approved.txt
Comment (Shane32) [inline src/GraphQL.ApiTests/net6/GraphQL.approved.txt:2433]: Remove?

**Comment by Shane32** [comment]
@sungam3r do you think Message should be the default mode?

We could have Extension be the default, and write code within the ExposeExceptionStackTrace property to set it equal to Message

File: src/GraphQL/Execution/ErrorInfoProviderOptions.cs
Comment (Shane32) [inline src/GraphQL/Execution/ErrorInfoProviderOptions.cs:17]:
```suggestion
            set
            {
                ExposeExceptionDetails = value;
                if (value)
                    ExposeExceptionDetailsMode = Message;
            }
```
Just an idea.....

File: src/GraphQL.ApiTests/net6/GraphQL.approved.txt
Comment (sungam3r) [reply src/GraphQL.ApiTests/net6/GraphQL.approved.txt:2433]: ooops, leak of local experiments

Comment (sungam3r) [reply src/GraphQL.ApiTests/net6/GraphQL.approved.txt]: done

File: src/GraphQL/Execution/ErrorInfoProviderOptions.cs
Comment (sungam3r) [reply src/GraphQL/Execution/ErrorInfoProviderOptions.cs:17]: > do you think Message should be the default mode?

for v5.4 yes, for v7 I would like to switch it to Extensions and remove `ExposeExceptionStackTrace`

Comment (Shane32) [reply src/GraphQL/Execution/ErrorInfoProviderOptions.cs:17]: I don't see any good reason to remove the member now, immediately before v7 is released.  It can be a deprecated member for v7.

I also would not switch the default of the new property between versions.

I would do as I posted above -- setting the old property retains the old behavior, while setting the new property defaults to the new behavior.  There is no breaking change for 5.4 and no breaking change when 7 is released.  The [Obsolete] comment needs to be updated to reflect this of course.

Comment (sungam3r) [reply src/GraphQL/Execution/ErrorInfoProviderOptions.cs:17]: ExposeExceptionDetailsMode defaults to `Message`. Why to reset it to Message again into setter ?

Comment (sungam3r) [reply src/GraphQL/Execution/ErrorInfoProviderOptions.cs:17]: ExposeExceptionStackTrace is the same thing as new property. It is just a way to not renaming it and introduce breaking changes. I don't want to bring some special behavior into setter.

Comment (sungam3r) [reply src/GraphQL/Execution/ErrorInfoProviderOptions.cs:17]: The goal is to eventually remove `ExposeExceptionStackTrace` and _maybe_ change default mode from Message to Extensions. 

I'm done for today, too tired to think more.

Comment (Shane32) [reply src/GraphQL/Execution/ErrorInfoProviderOptions.cs:17]: > ExposeExceptionDetailsMode defaults to Message. Why to reset it to Message again into setter ?

You said you wanted Extension to be the default, which I agree with.

> I don't want to bring some special behavior into setter.

Then you can't release this in 5.4.  It must be v7.  If you release it in 5.4 then it will cause a breaking change that is not marked by an obsolete flag in v7.  There is no reason to do so with v7 so near to being released.

Either that or Message should continue to be the default after v7 is released.

Or you go with my 3rd option, where using the new option uses the new default, and using the old option uses the old default.

Comment (Shane32) [reply src/GraphQL/Execution/ErrorInfoProviderOptions.cs:17]: I really don't see why not to do as I suggested.  It allows for non-API-breaking behavior in both v5.4 and v7, while having the new default value of 'Extension' for all new code.  Then we can remove the property in v8.

>  I don't want to bring some special behavior into setter.

Nor do I, but in this case, it is clearly the best choice for the scenario.  (in my opinion)

Comment (Shane32) [reply src/GraphQL/Execution/ErrorInfoProviderOptions.cs:17]: > The goal is to eventually remove `ExposeExceptionStackTrace` and _maybe_ change default mode from Message to Extensions.

If there is no goal to either change the default from Message or to remove the old property in v7, then we can merge as-is.  We should think about it and decide one way or the other.

Comment (sungam3r) [reply src/GraphQL/Execution/ErrorInfoProviderOptions.cs]: I was busy again. Now I looked at the code again and realized that the difference was not fundamental to me at all. As a result, the old property will eventually be deleted, and now I will not use it. I applied suggestion.

File: src/GraphQL/Execution/ErrorInfoProviderOptions.cs
Comment (Shane32) [inline src/GraphQL/Execution/ErrorInfoProviderOptions.cs:32]:
```suggestion
        public ExposeExceptionDetailsMode ExposeExceptionDetailsMode { get; set; } = ExposeExceptionDetailsMode.Extensions;
```

I thought we both agreed this would be better?  Note that with the code in the deprecated `ExposeExceptionStackTrace` property, this is a non-breaking change.

Comment (sungam3r) [reply src/GraphQL/Execution/ErrorInfoProviderOptions.cs:32]: ok

**Comment by sungam3r** [comment]
I retargeted to `develop` and added docs.

**Review by Shane32 — APPROVED** [review]
API approvals need to be updated, by the way.

File: src/GraphQL.ApiTests/net6/GraphQL.approved.txt
Comment (sungam3r) [inline src/GraphQL.ApiTests/net6/GraphQL.approved.txt:1159]:
```suggestion
        public ExecutionStrategyRegistration(GraphQL.Execution.IExecutionStrategy Strategy, GraphQLParser.AST.OperationType Operation) { }
        protected virtual System.Type EqualityContract { get; }
```

File: src/GraphQL.ApiTests/netstandard20+netstandard21/GraphQL.approved.txt
Comment (sungam3r) [inline src/GraphQL.ApiTests/netstandard20+netstandard21/GraphQL.approved.txt:1159]:
```suggestion
        public ExecutionStrategyRegistration(GraphQL.Execution.IExecutionStrategy Strategy, GraphQLParser.AST.OperationType Operation) { }
        protected virtual System.Type EqualityContract { get; }
```

**Comment by sungam3r** [comment]
I hope @stakx will accept fix in API approver to not bother anymore with this tiny property.

File: docs2/site/docs/migrations/migration7.md
Comment (sungam3r) [inline docs2/site/docs/migrations/migration7.md:231]: @Shane32 Please review docs for grammar flaws.

Comment (Shane32) [reply docs2/site/docs/migrations/migration7.md:231]: I looked through it briefly, but I thought perhaps I would take a pass over the entire document before we release 7.0 -- I was going to do that today (my time), after you merge this PR.  Perhaps you can review my changes tonight (your time) and if it's okay with you, I'll release 7.0 after that (tonight my time).

Mainly I'm thinking of re-ordering migration notes, either to make them flow better, or to draw attention to important notes.  Probably fairly light changes.

Comment (sungam3r) [reply docs2/site/docs/migrations/migration7.md:231]: So I merge this one.

**Pull request merged**
